### ISSUE TITLE
Android 13 patch file causes build errors

### ISSUE THREAD

**Issue by petefoth** [lineageos4microg/docker-lineage-cicd#353]
I am trying to make a 20.0 build using the the `lineageos4microg/docker-lineage-cicd:lineage-20` image with `-e "SIGNATURE_SPOOFING=restricted" `

The build failed with two errors, when compiling the patched file `frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java`.

The log showed the following
```
frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java:1595: error: cannot find symbol
        Bundle metaData = p.getMetaData();
        ^
  symbol:   class Bundle
  location: class ComputerEngine
frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java:1651: error: cannot find symbol
                    mPermissionManager.getGrantedPermissions(ps.name, userId)
                                                               ^
  symbol:   variable name
  location: variable ps of type PackageStateInternal
2 errors
```
These lines of code are inserted by the `android_frameworks_base-Android13.patch` file 
- in [lines 91-97](https://github.com/lineageos4microg/docker-lineage-cicd/blob/7fee64157f2b4d8c422cfdae6ff180044b3a5f56/src/signature_spoofing_patches/android_frameworks_base-Android13.patch#L91)
- in [lines 126-130](https://github.com/lineageos4microg/docker-lineage-cicd/blob/7fee64157f2b4d8c422cfdae6ff180044b3a5f56/src/signature_spoofing_patches/android_frameworks_base-Android13.patch#L126)

I have looked at the source file, but I don't know enough Java to understand why the compilation is failing or how to fix it

**Comment by Exodusnick** [comment]
[lineage-20.0-20221025-microG-signed-herolte.log](https://github.com/lineageos4microg/docker-lineage-cicd/files/9862214/lineage-20.0-20221025-microG-signed-herolte.log)
For me works see log

**Comment by petefoth** [comment]
> For me works see log
Curious!

Are you building with `SIGNATURE_SPOOFING=restricted`?

And the log does not mention the source file in question (`frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java` nor the  component `server/pm/` component of which it is part. 

Does your build contain microG GmsCore, and does it run successfully?

**Comment by Exodusnick** [comment]
Yes build with restricted and microg GmsCore.

**Comment by petefoth** [comment]
Thanks for your help.

Sorry - some more questions. I'm trying to work out why it works for you and not for me, and I **really** want to make a working 20.0 build for my device

1. Were the signature spoofing patches applied successfully? I have this in the build output
```
>> [Wed Oct 19 16:21:43 UTC 2022] Applying the restricted signature spoofing patch (based on android_frameworks_base-Android13.patch) to frameworks/base
patching file core/api/current.txt
patching file core/res/AndroidManifest.xml
patching file core/res/res/values/config.xml
patching file core/res/res/values/strings.xml
patching file services/core/java/com/android/server/pm/ComputerEngine.java
>
``` 
2. Does the content of the file `frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java` match the code shown above at (or around) lines 1595 and 1651?
3. Does Settings | System | microG Settings **run** OK? (My first 20.0 build completed, but microG Settings would not run, because I had forgotten to include  the `SIGNATURE_SPOOFING=restricted` in my build command)
4. What system are you running? My cloud instance was on ubuntu 18.04. (I think that **shouldn't** matter, but I am really clutching at straws now)
5. Were you building on a clean source tree?

Thanks again!

**Comment by petefoth** [comment]
I've tried a complete build, this time on a Ubuntu 20.04 instance.

**Comment by skulko** [comment]
Building the LineageOS sources with this patch, i got the same issue (build fails at 99%). LOS 19.1 with corresponding patch works.

[https://hastebin.com/tekirefora.coffeescript](log)

**Comment by petefoth** [comment]
> Building the LineageOS sources with this patch, i got the same issue (build fails at 99%).

Did you build using Docker?

What was your `docker run` command?

**Comment by skulko** [comment]
> Did you build using Docker?

No, I'm using the official LOS sources, no Docker. Buiild command was 'mka target-files-package otatools'

**Comment by petefoth** [comment]
Thanks

> No, I'm using the official LOS sources,

Is there any information in the LineageOS documentation about how to include MicroG,  F-Droid etc in the build? (With Docker, I just set `WITH_GMS=true`, and include `lineageos4microg/android_vendor_partner_gms` in my manifest.)

And do you apply the signature spoofing patches manually, or do you use a build script?

**Comment by SteadyQuad** [comment]
I attempted to fix the 2 compilation issues caused by the Android 13 frameworks patch, and together with @petefoth 's init.sh patch to generate some extra keys the ROM now technically builds.  

However that's about all the good news: the resulting ROM won't boot on my device, doesn't get past the LineageOS boot animation, and I don't even get to look at logcat.

Since the fixes don't seem to yield a working ROM, I won't submit a PR for these fixes at this point, but for who wants to take a look: https://github.com/SteadyQuad/docker-lineage-cicd/tree/lineage-20_sq1

**Comment by skulko** [comment]
> Do you apply the signature spoofing patches manually, or do you use a build script?

Yes and no, I have script that applies the different patches automatically. Have a look here: https://github.com/skulko/android_device_sony_xz2c/tree/lineage-19.1-microg/microg (patch/apply.sh). The required apks for microg, I add to the device.mk. (NOTICE: I did not write the script, just adapted it a little bit).

I don't like docker because it takes a lot of disk space with packages I already have installed.

**Comment by SteadyQuad** [comment]
> > Do you apply the signature spoofing patches manually, or do you use a build script?
> 
> Yes and no, I have script that applies the different patches automatically. Have a look here: https://github.com/skulko/android_device_sony_xz2c/tree/lineage-19.1-microg/microg (patch/apply.sh). The required apks for microg, I add to the device.mk. (NOTICE: I did not write the script, just adapted it a little bit).
> 

I comfused as to how this would build without issue, e.g. [this line](https://github.com/skulko/android_device_sony_xz2c/blob/c59b251f1a9bff28b6f45ab2b9ff030b5e14878d/microg/patch/frameworks_base/0001-Signature-Spoofing-Patch-13-LOS20.patch#L93) uses the Bundle class, which is not imported to ComputerEngine.java after applying the patch, and should give an error during compilation (as seen by others in the comments for this issue). Only after adding an import statement to ComputerEngine.java for Bundle it it builds for me.

**Comment by SteadyQuad** [comment]
I did another LOS20 build with signature patch (including my 2 fixes that let it build), but this time without including GMS (`WITH_GMS=false`). Now the ROM runs on my device, and after installing GMS core separately through F-Droid, I get this microG self-check, with a lot of checks missing (the 2 checked items are the only ones in the list). Not sure how to troubleshoot this efficiently, any advice?

![Screenshot_20221107-113542_microG Services Core](https://user-images.githubusercontent.com/11686811/200312872-b7702e14-5565-4f78-ae05-1213f9443937.png)

**Comment by skulko** [comment]
>  Only after adding an import statement to ComputerEngine.java for Bundle it it builds for me.

Why not add the import statement to the patch here? There seems to be no such line in the original LOS sources, so it would be useful. As mentioned, I took the patch for Android 12 from here, applied it on the cloned LOS sources (19.1), and it is working nicely. I only mentioned my issue here because I had the same error message as @petefoth .

**Comment by SteadyQuad** [comment]
> Why not add the import statement to the patch here? There seems to be no such line in the original LOS sources, so it would be useful. As mentioned, I took the patch for Android 12 from here, applied it on the cloned LOS sources (19.1), and it is working nicely. I only mentioned my issue here because I had the same error message as @petefoth .

Adding an import would not be a problem, the mystery to me is that you don't seem to have a problem building LOS20 with the signature spoofing patch as it is in this repo, while I and @petefoth run into two compilation errors with the same patch.

Could it be that you're buidling with an adapted version of the Android 13 signature spoofing patch from [here](https://github.com/danielk43/docker-lineage-cicd/tree/lineage-20-8890) instead of the patch from this lineageos4microg/docker-lineage-cicd repo?

That version seems to be re-written/fixed to actually compile out of the box.

**Comment by SteadyQuad** [comment]
FWIW: I was abe to build a working ROM with GMS core included by using work from repos of [danielk43](https://github.com/danielk43), including their version of the [Android 13 signature spoofing patch](https://github.com/danielk43/docker-lineage-cicd/raw/lineage-20-8890/src/signature_spoofing_patches/android_frameworks_base-Android13.patch) and including `danielk43/android_vendor_partner_gms` instead of `lineageos4microg/android_vendor_partner_gms` in a local manifest.

**Comment by danielk43** [comment]
Hiya, @Exodusnick made me aware of this thread.. just a heads-up I think we have solved all the remaining A13 problems now. Most recently signing Bluetooth.apk with the new bt key.

see [this](https://github.com/lineageos4microg/docker-lineage-cicd/pull/359) PR

**Comment by Exodusnick** [comment]
@danielk43 Thank you new build from 20.11.2022 Bluetooth now works. Just like other things 👍

**Comment by petefoth** [comment]
> I think we have solved all the remaining A13 problems now. Most recently signing Bluetooth.apk with the new bt key.

Great stuff. Do you have any idea when the changes will be merged into the `lineage-20` branch?

**Comment by SteadyQuad** [comment]
> Hiya, @Exodusnick made me aware of this thread.. just a heads-up I think we have solved all the remaining A13 problems now. Most recently signing Bluetooth.apk with the new bt key.

Great! Do you think you have any changes in danielk43/android_vendor_partner_gms that are relevant to getting A13 to work? If I remember I was not able to boot a LOS20 ROM build with included lineageos4microg/android_vendor_partner_gms, but including yours instead worked ok.

**Comment by danielk43** [comment]
@petefoth Best guess is they are waiting for the official release, but you could try my docker image before that. Feel free to use `danielk43/docker-lineage-cicd:los20_fixes` as the build container.

@SteadyQuad 2 changes to GmsCore - one related to A13 and another required for the latest microG. Actually there are 2 PRs open so whichever one ends up merged, there is a lot of awareness now.

Optionally, the local manifests from [here](https://github.com/danielk43/docker-lineage-cicd/tree/lineage-20-8890/src/8890_manifests) (except 8890) will build with some additional hardening, Quad9 DNS, and Bromite browser/webview

**Comment by SteadyQuad** [comment]
FYI I think your bluetooth certificate hack in: https://github.com/lineageos4microg/docker-lineage-cicd/pull/359/commits/4b103386b52a91f962197c2e9c7eb9ffac8eb4ab is no longer necessary for LOS20 builds, they're using [their own fork](https://review.lineageos.org/c/LineageOS/android/+/344543) of bluetooth module now, which does [not use a local certificate](https://review.lineageos.org/c/LineageOS/android_packages_modules_Bluetooth/+/344542) anymore.

**Comment by danielk43** [comment]
Yea, thanks I saw also.. testing with repopicks instead of `cp`

**Comment by petefoth** [comment]
> Best guess is they are waiting for the official release, but you could try my docker image before that. Feel free to use `danielk43/docker-lineage-cicd:los20_fixes` as the build container.

Thanks. I tried that, and the builds succeeded, but they failed to boot because of the problem reported in #360. 

There's a fix for that issue implemented in a [Pull Request](https://github.com/lineageos4microg/android_vendor_partner_gms/pull/15) in [android_vendor_partner_gms](https://github.com/lineageos4microg/android_vendor_partner_gms)

In the mean time, there's a workaround [described in the XDA Forum thread](https://forum.xda-developers.com/t/lineageos-for-microg.3700997/post-87757075) which works, but when I use that, with my builds made using your docker image, dirty flash and TWRP backup restore are broken. Rather than try to debug all that, I'm going to wait until all these changes get merged (in this repo and `android_vendor_partner_gms`) before I make any builds for public consumption :)

Thanks again

**Comment by petefoth** [comment]
> you can pin his or my fork of `android_vendor_partner_gms` in you local manifest. the issue is addressed in both

Thanks. There's also this issue of whether or not the bluetooth certificate fix/hack mentioned above is needed or not. My aim is to have builds that are easy to reproduce and to update, and that is easiest if I can build using the main line of development. For me, keeping track of which workarounds and forks I am using, causes too much hassle. 

I do appreciate the efforts of everyone who is working on this, but I'm not in a great rush, so I'm happy to wait for these changes to get merged in the master or lineage-20 branches of the official repos (so long as that doesn't take too long ;) )

**Comment by danielk43** [comment]
the bt fixes were merged and already removed from the pr/build

**Comment by petefoth** [comment]
I can confirm this bug was fixed by [this commit](https://github.com/lineageos4microg/docker-lineage-cicd/commit/1380af9848a735209d4fea0dd51d9eb33e7842b8) in the `lineage-20` branch. 

I'll close this issue now

**Comment by Exodusnick** [comment]
@danielk43 Please get in touch I would like to clarify something.
